### PR TITLE
td-payload: handle `retry` result of mapping GPA

### PR DESCRIPTION
**Pull request by gaojiaqi7**
Fix: https://github.com/confidential-containers/td-shim/issues/498

### CODE REVIEW

File: td-payload/src/arch/x86_64/dma.rs [hunk 1]
```diff
@@ -3,16 +3,28 @@
 // SPDX-License-Identifier: BSD-2-Clause-Patent
 
 use crate::mm::SIZE_4K;
-use tdx_tdcall::tdx;
+use tdx_tdcall::{tdx, TdVmcallError};
 
 use super::paging::{clear_shared_bit, set_shared_bit};
 
+const MAP_GPA_MAX_RETRIES: u8 = 3;
```
Comment (jyao1) [inline td-payload/src/arch/x86_64/dma.rs:10]: why retry 3 times?

**Comment by gaojiaqi7** [comment]
Screenshot 2023-03-20 212728

**Comment by gaojiaqi7** [comment]
Answer from KVM team:

> 1. Under what circumstances VMM may return this result?

When the specified range is too large. VMM can't complete the request in practical time. e.g. hundreds of GB.

It's up to VMM implementation.


> 2. The spec says guest TD must retry this operation when `TDG.VP.VMCALL_RETRY` is returned. Should guest TD set any maximum retries or timeout? or it should try again unconditionally, no matter how many times it has already tried?

Retry unconditionally. GHCI spec doesn't have any guide on the retry because it's attack surface for (malicious) VMM against guest TD.

If you want guest to implement guard, it's totally up to you how to protect it. probably it's heuristic protection.

**Comment by mxu9** [comment]
It looks good to me.

**Review by jyao1 — APPROVED** [review]

**Pull request merged**